Add tests and implementation for connections (use-case 1.xii)

dladd opened the pull request:
Addresses issue #88

dladd committed:
add initial tests for connections

dladd committed:
add CMake test interface for connections

dladd committed:
add specific tests for use-cases 1.xii.a.1-2

nickerso commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,281 @@
+      "<component name=\"component1\">"
+        "<variable name=\"variable1\"/>"
+      "</component>"
+      "<variable name=\"variable2\"/>"
```
I wouldn't expect to see this here.

nickerso commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,281 @@
+    v2.setInterfaceType(libcellml::Variable::INTERFACE_TYPE_PUBLIC);
+    v3.setInterfaceType(libcellml::Variable::INTERFACE_TYPE_PUBLIC);
+
+    std::string a = parent.serialise(libcellml::FORMAT_XML);
```
should be `m.serialise(...)` :)

dladd committed:
bug fixes for initial tests

dladd committed:
further bugfixes for initial connection tests

dladd committed:
add duplicate eqiv. var. test; bugfix for reciprocal tests

dladd committed:
begin implementing equivalent variable side of connection code

dladd commented on src/variable.cpp:
```diff
@@ -77,6 +87,36 @@ void Variable::swap(Variable &rhs)
+void Variable::addEquivalence(const VariablePtr &equivalentVariable)
+{
+    setEquivalentTo(equivalentVariable);
+    VariablePtr *pV = new VariablePtr(this);
```
For anyone that happens to be looking through this now, I definitely plan to sort out a safer and more C++11 way to do this tomorrow. I was just using this to check that the rest of the current code passes the variable-side connections tests. 

Of course, any suggestions are always welcome :-)

nickerso replied:
I'm starting to wonder if, given the reciprocal nature of this method, perhaps it is better implemented as a static method on the variable class that takes two `VariablePtr` arguments?

dladd replied:
A static method does indeed seem to get around the issue nicely. It also takes some of the possible ambiguity about reciprocity for anyone not digging into the documentation (I would expect reciprocity out of `addEquivalence(v1,v2)` but not necessarily out of `v1.addEquivalence(v2)`)

dladd committed:
change addEquivalence() to static method

dladd committed:
set variable parent components; test

dladd committed:
add two connection test; test bugfixes

dladd committed:
working state for non-encapsulated connection serialisation

dladd committed:
move location of build variable map step to capture encapsulated components

dladd committed:
add more complex connections tests (nineVariablesTenConnections, threeMapVariablesConnectionOneDuplicate)

dladd committed:
remove redundant reciprocal component check

dladd committed:
some doc and formatting tidy-up for new features.

dladd committed:
modify mEquivalentVariables to use weak_ptrs: avoids memory leaks from cyclical shared_ptrs

dladd committed:
minor bits of doc/formatting tidy-up

dladd commented:
This now passes all tests and proposed use-cases. A few notes:
- Per discussion at the end of #85 , I've simplified the connection serialisation from the original object model proposal with an explicit private `VariableMap` class. Instead, the same functionality is implemented using vectors of c++ `pairs` in `Model::doSerialisation`. This is also more consistent with the way the `ImportMap` is currently constructed.
- `addEquivalence(v1,v2)` is now implemented as a static `Variable` function. This will hopefully make the reciprocity of this method more obvious to the user than the `v1.addEquivalence(v2)` proposal. 

~~Another issue I noticed is that we haven't yet properly tested serialisation of variables inside imported components. If there are no objections, I'll add that as a separate bugfix request pending the review of this.~~

dladd committed:
add imported component connection test

dladd commented:
Apologies- please ignore the above comment re: imported component variables. @nickerso confirmed for me that we don't expect variables to be serialised from imported components. 

I've also now added a passing test for a connection between imported and non-imported components.

nickerso commented on src/model.cpp:
```diff
@@ -51,10 +52,22 @@ Model& Model::operator=(Model m)
+    // ComponentMap
+    typedef std::pair <Component*, Component*> ComponentPair;
+    typedef std::vector<ComponentPair>::const_iterator ComponentMapIterator;
+    typedef std::vector<ComponentPair> ComponentMap;
```
I'd probably change the code slightly for these to make it a bit cleaner, maybe :)

``` c
typedef std::pair<Component*, Component*> ComponentPair;
typedef std::vector<ComponentPair> ComponentMap;
typedef ComponentMap::const_iterator ComponentMapIterator;
```

dladd committed:
tidy-up variable/component map typedefs

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
         repr += serialiseUnits(format);
 
-        // Serialize components of the model, imported components have already been dealt with at this point.
+        // Serialise components of the model, imported components have already been dealt with at this point.
```
Nitpicking: sometimes comments finish with `.`, sometimes not...

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
+            ComponentPtr component = getComponent(i);
+            for (size_t j = 0; j < component->variableCount(); ++j) {
+                VariablePtr variable = component->getVariable(j);
+                if (variable->equivalentVariableCount() > 0) {
```
I would drop the `> 0` in the test, but that's just me.

jonc125 replied:
I'd keep it here, as you're comparing a count to some value - the method doesn't return a boolean. So this way is semantically clearer.

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
+                        VariablePair reciprocalVariablePair = std::make_pair(equivalentVariable, variable);
+                        bool pairFound = false;
+                        for (VariableMapIterator iter = variableMap.begin(); iter < variableMap.end(); ++iter) {
+                            if (*iter == variablePair || *iter == reciprocalVariablePair) {
```
I haven't checked the coding style, but personally I prefer to make it (very) clear what we are testing. In other words, I would have:

```
...
if ((*iter == variablePair) || (*iter == reciprocalVariablePair)) {
...
```

Same elsewhere in the code (e.g. line 185 here).

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
+            }
+            // Serialise out the new connection
+            std::string connection = "<connection><map_components";
+            if (currentComponent1 != nullptr) {
```
As for the `> 0` above, I would drop the `!= nullptr`, but again that's just me. Similarly, I wouldn't use curly brackets here since the IF statement is a single-line statement, but then again the Google C++ style guide allows it...

jonc125 replied:
In this case I think treating the pointer as a boolean is actually a more normal approach to testing for null with smart pointers, although the pointers here are raw pointers so less clear.

And I'd keep the curly brackets - avoids future mistakes if the if body becomes multi statement :)

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
+            // Serialise out the new connection
+            std::string connection = "<connection><map_components";
+            if (currentComponent1 != nullptr) {
+                connection += " component_1=\"" + currentComponent1->getName() + "\"";
```
I wouldn't have spaces around `+` (same elsewhere in the code).

hsorby replied:
Always spaces around operators!

agarny commented on src/model.cpp:
```diff
@@ -127,9 +140,87 @@ std::string Model::doSerialisation(FORMATS format) const
+            int componentMapIndex2 = componentMapIndex1 + 1;
+            ComponentMapIterator duplicateIterPair = iterPair;
+            ++duplicateIterPair;
+            for (ComponentMapIterator iterPair2 = duplicateIterPair; iterPair2 < componentMap.end(); ++iterPair2) {
```
Can't `duplicateIterPair` be dropped and have `... iterPair2 = iterPair+1; ...` instead?

agarny commented on src/variable.cpp:
```diff
@@ -31,17 +33,27 @@ namespace libcellml {
+std::vector<VariableWeakPtr>::iterator Variable::VariableImpl::findEquivalentVariable(const VariablePtr &equivalentVariable)
+{
+    return std::find_if(mEquivalentVariables.begin(), mEquivalentVariables.end(),
+                        [=](VariableWeakPtr variableWeak) -> bool {return equivalentVariable == variableWeak.lock();});
```
I would put a space after `{` and before `}`, i.e.

```
[=](VariableWeakPtr variableWeak) -> bool { return equivalentVariable == variableWeak.lock(); });
```

agarny commented on src/variable.cpp:
```diff
@@ -77,6 +90,37 @@ void Variable::swap(Variable &rhs)
+bool Variable::hasEquivalentVariable(const VariablePtr &equivalentVariable)
+{
+    auto result = mPimpl->findEquivalentVariable(equivalentVariable);
+    return result != mPimpl->mEquivalentVariables.end();
```
Why not simply:

```
return mPimpl->findEquivalentVariable(equivalentVariable) != mPimpl->mEquivalentVariables.end(); 
```

?

hsorby replied:
Line width issue? I think originally this was a hangover from debugging a problem.

dladd replied:
I've swapped in @agarny 's suggestion but you are right, this does violate the Google style guide on [max line lengths of 80 characters](https://google.github.io/styleguide/cppguide.html#Line_Length). We do appear to be +80 characters in many places though so it might be worth deciding if we want to rigidly follow Google on this or add an exception to our coding style.

agarny commented on tests/CMakeLists.txt:
```diff
@@ -18,6 +18,7 @@ include(component/tests.cmake)
 include(coverage/tests.cmake)
 include(units/tests.cmake)
 include(variable/tests.cmake)
+include(connection/tests.cmake)
```
Might want to order those includes?

agarny commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,546 @@
+    libcellml::VariablePtr v1 = std::make_shared<libcellml::Variable>();
+    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
+    libcellml::Variable::addEquivalence(v1,v2);
+    EXPECT_EQ(v2,v1->getEquivalentVariable(0));
```
Missing space before `v1` (same for `v2` at lines 24, 31 and 32). (Same elsewhere in the code.)

agarny commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,546 @@
+    libcellml::VariablePtr v3 = std::make_shared<libcellml::Variable>();
+    libcellml::Variable::addEquivalence(v1,v2);
+    libcellml::Variable::addEquivalence(v1,v3);
+    size_t e = 2;
```
Missing `const`? (Same elsewhere in this code.)

agarny commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,546 @@
+    libcellml::VariablePtr v21 = std::make_shared<libcellml::Variable>();
+    v21->setName("variable21");
+    libcellml::VariablePtr v22 = std::make_shared<libcellml::Variable>();
+    v22->setName("variable22");
```
I would personally declare everything and then set the names.

Otherwise, I like it that there are empty lines here and there in this test (in fact, I would have added a few more). I find it easier to read the code... (not only the tests, but the code in general)

agarny commented on tests/connection/connection.cpp:
```diff
@@ -0,0 +1,546 @@
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/1.2#\">"
+
+          "<component name=\"parent_component\">"
```
Might simply want to call it `parent`? (Same elsewhere in the code.)

agarny commented:
Quite a bit of nitpicking, sorry...

agarny commented:
With regards to issue #88, I think I am happy with this PR. I think because there is quite a bit of code to review, so I have just had a very quick glance at it, but it passes the tests, so...

dladd committed:
end sentence-style comments with punctuation

dladd committed:
separate ||/&& conditional arguments with ()

dladd committed:
remove unnecessary iterator

dladd committed:
adjust spacing on find_if() statements

dladd committed:
simplify hasEquivalentVariable()

dladd committed:
clean up connection tests per comments on pull 89

dladd committed:
remove instances of !=nullptr in conditionals

dladd commented:
Thanks all- this should now address the above comments.

agarny commented on src/component.cpp:
```diff
@@ -37,8 +37,8 @@ struct Component::ComponentImpl
 {
-    return std::find_if(mVariables.begin(), mVariables.end(),
-                        [=](const VariablePtr& v) -> bool {return v->getName() == name;});
+    return std::find_if (mVariables.begin(), mVariables.end(),
```
Just for information (i.e. don't make it delay the merge!): `std::find_if()` is a function, not a an `if` statement as such, so I wouldn't expect a space to be 'needed.

dladd replied:
Fair enough- I didn't see an instance of find_if in the Google style guide but did see it on [cplusplus.com](http://www.cplusplus.com/reference/algorithm/find_if/) with spacing. But that makes sense and my Stroustrup book has it without the spacing.

agarny commented:
Fine by me.

dladd committed:
fix spacing for find_if
